appIcons.js

```diff
@@ -542,6 +574,199 @@ const MyAppIcon = new Lang.Class({
             this._dots.queue_repaint();
     },
 
+    _enableBacklight: function() {
+        let pixBuf = this._getIconPixBuf();
+
+        // Fallback
+        if (pixBuf === null) {
+            this._iconContainer.set_style(
+                'border-radius: 5px;' +
+                'background-gradient-direction: vertical;' +
+                'background-gradient-start: #e0e0e0;' +
+                'background-gradient-end: darkgray;'
+            );
+            
+            this._dot.set_style(
+                'height: 0px;' +
+                'border: 0px solid white;' +
+                'background-color: gray;'
+            );
+
+            return;
+        }
+        
+        let colorPallete = this._calculateColorPalette(pixBuf);
+
+        this._iconContainer.set_style(
+            'border-radius: 5px;' +
+            'background-gradient-direction: vertical;' +
+            'background-gradient-start: ' + colorPallete.original + ';' +
+            'background-gradient-end: ' +  colorPallete.darker + ';'
+        );
+
+        this._dot.set_style(
+            'height: 0px;' +
+            'border: 0px solid ' + colorPallete.lighter  + ';' +
+            'background-color: ' + colorPallete.darker + ';'
+        );
+    },
+
+    _toggleBacklight: function () {
+        this._glossyBackground();
+        this._updateRunningStyle();
+        
+        if (this._dots)
+            this._dots.queue_repaint();
+    },
+
+    _disableBacklight: function() {
+        this._iconContainer.set_style(null);
+        this._dot.set_style(null);
+    },
+    
+    /**
+     * Try to get the pixel buffer for the current icon, if not fail gracefully     
+     */
+    _getIconPixBuf: function() {
+        let iconTexture = this.app.create_icon_texture(16);
+
+        if (themeLoader === null) {
+            let ifaceSettings = new Gio.Settings({ schema: "org.gnome.desktop.interface" });
+
+            themeLoader = new Gtk.IconTheme(),
+            themeLoader.set_custom_theme(ifaceSettings.get_string('icon-theme')); // Make sure the correct theme is loaded
+        }        
+
+        // Unable to load the icon texture, use fallback
+        if (iconTexture instanceof St.Icon === false) {
+            return null;
+        }
+
+        iconTexture = iconTexture.get_gicon();       
+
+        // Unable to load the icon texture, use fallback
+        if (iconTexture === null) {
+            return null;
+        }
+
+        if (iconTexture instanceof Gio.FileIcon) {
+            // Use GdkPixBuf to load the pixel buffer from the provided file path
+            return GdkPixbuf.Pixbuf.new_from_file(iconTexture.get_file().get_path());
+        }
+        
+        // Get the pixel buffer from the icon theme
+        return themeLoader.load_icon(iconTexture.get_names()[0], 24, 0);
```

Comment:
I agree, after adding cache I think that would have minimal impact, and gives us better results.